server/src/main/java/com/cloud/storage/StorageManagerImpl.java

```diff
@@ -684,12 +685,19 @@ public boolean stop() {
         return true;
     }
 
-    private DataStore createLocalStorage(Map<String, Object> poolInfos) throws ConnectionException{
+    protected DataStore createLocalStorage(Map<String, Object> poolInfos) throws ConnectionException{
         Object existingUuid = poolInfos.get("uuid");
         if( existingUuid == null ){
             poolInfos.put("uuid", UUID.randomUUID().toString());
         }
-        String hostAddress = poolInfos.get("host").toString();
+        String hostAddress = poolInfos.get("host") == null ? null : poolInfos.get("host").toString();
+        if (StringUtils.isEmpty(hostAddress)) {
+            throw new InvalidParameterValueException("Invalid host provided");
+        }
+        String hostPath = poolInfos.get("hostPath") == null ? null : poolInfos.get("hostPath").toString();
+        if (StringUtils.isEmpty(hostPath)) {
+            throw new InvalidParameterValueException("Invalid path provided");
+        }
```

Comment:
can these be two extra methods for readability?

Reply:
this looks more readable to me  :-D

Reply:
I think we should have more discipline to keep methods simple. Our code should be readable by humans *and* computers.

Reply:
on the other hand if there are too many methods and method hierarchies, it is not readable by humans.
we need a tradeoff

Reply:
@DaanHoogland I've done the refactor. Do you approve the changes?

Reply:
yes @shwstppr 
@weizhouapache if that is the case we need more refactoring, not less. Organising the methods into logical groups and extract them into their own modules. The only trade-off is "Do we have enough time to do it" imnsho.